## Programa Ingenias+ Data Science

## Trabajando con Pandas

1) Importa las librerias requeridas: `Pandas` y `NumPy`.  

Una buena practica a la hora de escribir codigo es de comentar siempre que estoy haciendo (usando `#`) en la linea anterior. Por mas obvio que parezca lo que voy a hacer, puede ser que tenga que compartir el documento con otras personas y no sepan que quise hacer en cada linea. O que yo mismo lea el codigo tiempo despues y no recuerde porque lo hice.

In [ ]:
import pandas as pd
import numpy as np

2) Lee el archivo csv `life_expectancy_data` usando `pandas`. Guardalo en una variable llamada `expvida`.

In [64]:
# Leo los datos en un DataFrame
variable = pd.read_csv("archivo.csv")


In [21]:
import pandas as pd
expvida = pd.read_csv("archivo.csv")
expvida

Country  Year      Status  Life expectancy   Adult Mortality  \
0     Afghanistan  2015  Developing              65.0            263.0   
1     Afghanistan  2014  Developing              59.9            271.0   
2     Afghanistan  2013  Developing              59.9            268.0   
3     Afghanistan  2012  Developing              59.5            272.0   
4     Afghanistan  2011  Developing              59.2            275.0   
...           ...   ...         ...               ...              ...   
2933     Zimbabwe  2004  Developing              44.3            723.0   
2934     Zimbabwe  2003  Developing              44.5            715.0   
2935     Zimbabwe  2002  Developing              44.8             73.0   
2936     Zimbabwe  2001  Developing              45.3            686.0   
2937     Zimbabwe  2000  Developing              46.0            665.0   

      infant deaths  Alcohol  percentage expenditure  Hepatitis B  Measles   \
0                62     0.01               71.279624         65.0      1154   
1                64     0.01               73.523582         62.0       492   
2                66     0.01               73.219243         64.0       430   
3                69     0.01               78.184215         67.0      2787   
4                71     0.01                7.097109         68.0      3013   
...             ...      ...                     ...          ...       ...   
2933             27     4.36                0.000000         68.0        31   
2934             26     4.06                0.000000          7.0       998   
2935             25     4.43                0.000000         73.0       304   
2936             25     1.72                0.000000         76.0       529   
2937             24     1.68                0.000000         79.0      1483   

      ...  Polio  Total expenditure  Diphtheria    HIV/AIDS         GDP  \
0     ...    6.0               8.16         65.0        0.1  584.259210   
1     ...   58.0               8.18         62.0        0.1  612.696514   
2     ...   62.0               8.13         64.0        0.1  631.744976   
3     ...   67.0               8.52         67.0        0.1  669.959000   
4     ...   68.0               7.87         68.0        0.1   63.537231   
...   ...    ...                ...          ...        ...         ...   
2933  ...   67.0               7.13         65.0       33.6  454.366654   
2934  ...    7.0               6.52         68.0       36.7  453.351155   
2935  ...   73.0               6.53         71.0       39.8   57.348340   
2936  ...   76.0               6.16         75.0       42.1  548.587312   
2937  ...   78.0               7.10         78.0       43.5  547.358878   

      Population   thinness  1-19 years   thinness 5-9 years  \
0     33736494.0                   17.2                 17.3   
1       327582.0                   17.5                 17.5   
2     31731688.0                   17.7                 17.7   
3      3696958.0                   17.9                 18.0   
4      2978599.0                   18.2                 18.2   
...          ...                    ...                  ...   
2933  12777511.0                    9.4                  9.4   
2934  12633897.0                    9.8                  9.9   
2935    125525.0                    1.2                  1.3   
2936  12366165.0                    1.6                  1.7   
2937  12222251.0                   11.0                 11.2   

      Income composition of resources  Schooling  
0                               0.479       10.1  
1                               0.476       10.0  
2                               0.470        9.9  
3                               0.463        9.8  
4                               0.454        9.5  
...                               ...        ...  
2933                            0.407        9.2  
2934                            0.418        9.5  
2935                            0.427      

3) ¿Que tipo de estructura de datos contiene la variable `expvida`?

In [ ]:
# Obtengo el tipo de estructura de datos
type expvida

In [ ]:
pandas.core.frame.DataFrame

4) ¿Cuantas filas y columnas tiene `expvida`?

In [22]:
# Uso .shape para obtener cantidad de filas y columnas
expvida.shape

(2938, 22)

**Numero de filas**:2938

**Numero de columnas**: 29

5) ¿Cual es el nombre de las columnas contenidas en `expvida`?

In [23]:
# Obtengo el nombre de las columnas
expvida.columns

Index(['Country', 'Year', 'Status', 'Life expectancy ', 'Adult Mortality',
       'infant deaths', 'Alcohol', 'percentage expenditure', 'Hepatitis B',
       'Measles ', ' BMI ', 'under-five deaths ', 'Polio', 'Total expenditure',
       'Diphtheria ', ' HIV/AIDS', 'GDP', 'Population',
       ' thinness  1-19 years', ' thinness 5-9 years',
       'Income composition of resources', 'Schooling'],
      dtype='object')

6) Inspecciona las primeras 10 filas de `expvida`

In [24]:
# Inspecciono las primeras filas
expvida.head(10)

Country  Year      Status  Life expectancy   Adult Mortality  \
0  Afghanistan  2015  Developing              65.0            263.0   
1  Afghanistan  2014  Developing              59.9            271.0   
2  Afghanistan  2013  Developing              59.9            268.0   
3  Afghanistan  2012  Developing              59.5            272.0   
4  Afghanistan  2011  Developing              59.2            275.0   
5  Afghanistan  2010  Developing              58.8            279.0   
6  Afghanistan  2009  Developing              58.6            281.0   
7  Afghanistan  2008  Developing              58.1            287.0   
8  Afghanistan  2007  Developing              57.5            295.0   
9  Afghanistan  2006  Developing              57.3            295.0   

   infant deaths  Alcohol  percentage expenditure  Hepatitis B  Measles   ...  \
0             62     0.01               71.279624         65.0      1154  ...   
1             64     0.01               73.523582         62.0       492  ...   
2             66     0.01               73.219243         64.0       430  ...   
3             69     0.01               78.184215         67.0      2787  ...   
4             71     0.01                7.097109         68.0      3013  ...   
5             74     0.01               79.679367         66.0      1989  ...   
6             77     0.01               56.762217         63.0      2861  ...   
7             80     0.03               25.873925         64.0      1599  ...   
8             82     0.02               10.910156         63.0      1141  ...   
9             84     0.03               17.171518         64.0      1990  ...   

   Polio  Total expenditure  Diphtheria    HIV/AIDS         GDP  Population  \
0    6.0               8.16         65.0        0.1  584.259210  33736494.0   
1   58.0               8.18         62.0        0.1  612.696514    327582.0   
2   62.0               8.13         64.0        0.1  631.744976  31731688.0   
3   67.0               8.52         67.0        0.1  669.959000   3696958.0   
4   68.0               7.87         68.0        0.1   63.537231   2978599.0   
5   66.0               9.20         66.0        0.1  553.328940   2883167.0   
6   63.0               9.42         63.0        0.1  445.893298    284331.0   
7   64.0               8.33         64.0        0.1  373.361116   2729431.0   
8   63.0               6.73         63.0        0.1  369.835796  26616792.0   
9   58.0               7.43         58.0        0.1  272.563770   2589345.0   

    thinness  1-19 years   thinness 5-9 years  \
0                   17.2                 17.3   
1                   17.5                 17.5   
2                   17.7                 17.7   
3                   17.9                 18.0   
4                   18.2                 18.2   
5                   18.4                 18.4   
6                   18.6                 18.7   
7                   18.8                 18.9   
8                   19.0                 19.1   
9                   19.2                 19.3   

   Income composition of resources  Schooling  
0                            0.479       10.1  
1                            0.476       10.0  
2                            0.470        9.9  
3                            0.463        9.8  
4                            0.454        9.5  
5                            0.448        9.2  
6                            0.434        8.9  
7                            0.433        8.7  
8                            0.415        8.4  
9                            0.405        8.1  

[10 rows x 22 columns]

7) Y ahora las 10 ultimas.

In [27]:
# Inspecciono las ultimas filas
expvida.tail(10)

Country  Year      Status  Life expectancy   Adult Mortality  \
2928  Zimbabwe  2009  Developing              50.0            587.0   
2929  Zimbabwe  2008  Developing              48.2            632.0   
2930  Zimbabwe  2007  Developing              46.6             67.0   
2931  Zimbabwe  2006  Developing              45.4              7.0   
2932  Zimbabwe  2005  Developing              44.6            717.0   
2933  Zimbabwe  2004  Developing              44.3            723.0   
2934  Zimbabwe  2003  Developing              44.5            715.0   
2935  Zimbabwe  2002  Developing              44.8             73.0   
2936  Zimbabwe  2001  Developing              45.3            686.0   
2937  Zimbabwe  2000  Developing              46.0            665.0   

      infant deaths  Alcohol  percentage expenditure  Hepatitis B  Measles   \
2928             30     4.64                1.040021         73.0       853   
2929             30     3.56               20.843429         75.0         0   
2930             29     3.88               29.814566         72.0       242   
2931             28     4.57               34.262169         68.0       212   
2932             28     4.14                8.717409         65.0       420   
2933             27     4.36                0.000000         68.0        31   
2934             26     4.06                0.000000          7.0       998   
2935             25     4.43                0.000000         73.0       304   
2936             25     1.72                0.000000         76.0       529   
2937             24     1.68                0.000000         79.0      1483   

      ...  Polio  Total expenditure  Diphtheria    HIV/AIDS         GDP  \
2928  ...   69.0               6.26         73.0       18.1   65.824121   
2929  ...   75.0               4.96         75.0       20.5  325.678573   
2930  ...   73.0               4.47         73.0       23.7  396.998217   
2931  ...   71.0               5.12          7.0       26.8  414.796232   
2932  ...   69.0               6.44         68.0       30.3  444.765750   
2933  ...   67.0               7.13         65.0       33.6  454.366654   
2934  ...    7.0               6.52         68.0       36.7  453.351155   
2935  ...   73.0               6.53         71.0       39.8   57.348340   
2936  ...   76.0               6.16         75.0       42.1  548.587312   
2937  ...   78.0               7.10         78.0       43.5  547.358878   

      Population   thinness  1-19 years   thinness 5-9 years  \
2928   1381599.0                    7.5                  7.4   
2929  13558469.0                    7.8                  7.8   
2930   1332999.0                    8.2                  8.2   
2931  13124267.0                    8.6                  8.6   
2932    129432.0                    9.0                  9.0   
2933  12777511.0                    9.4                  9.4   
2934  12633897.0                    9.8                  9.9   
2935    125525.0                    1.2                  1.3   
2936  12366165.0                    1.6                  1.7   
2937  12222251.0                   11.0                 11.2   

      Income composition of resources  Schooling  
2928                            0.419        9.9  
2929                            0.421        9.7  
2930                            0.414        9.6  
2931                            0.408        9.5  
2932                            0.406        9.3  
2933                            0.407        9.2  
2934                            0.418        9.5  
2935                            0.427       10.0  
2936                            0.427        9.8  
2937                            0.434        9.8  

[10 rows x 22 columns]

8) ¿Que tipos de datos contiene cada una de las columnas de `expvida`?

In [30]:
# Tipo de datos en cada columna
expvida.dtypes

Country                             object
Year                                 int64
Status                              object
Life expectancy                    float64
Adult Mortality                    float64
infant deaths                        int64
Alcohol                            float64
percentage expenditure             float64
Hepatitis B                        float64
Measles                              int64
 BMI                               float64
under-five deaths                    int64
Polio                              float64
Total expenditure                  float64
Diphtheria                         float64
 HIV/AIDS                          float64
GDP                                float64
Population                         float64
 thinness  1-19 years              float64
 thinness 5-9 years                float64
Income composition of resources    float64
Schooling                          float64
dtype: object

9) ¿Hay valores faltantes en alguna de las columnas de `expvida`?

In [31]:
# Obtengo la cantidad de datos faltantes en cada columna
expvida.isnull().sum()

Country                              0
Year                                 0
Status                               0
Life expectancy                     10
Adult Mortality                     10
infant deaths                        0
Alcohol                            194
percentage expenditure               0
Hepatitis B                        553
Measles                              0
 BMI                                34
under-five deaths                    0
Polio                               19
Total expenditure                  226
Diphtheria                          19
 HIV/AIDS                            0
GDP                                448
Population                         652
 thinness  1-19 years               34
 thinness 5-9 years                 34
Income composition of resources    167
Schooling                          163
dtype: int64

10) Selecciona el maximo de datos faltantes. Calcula que porcentaje de los datos representan los datos faltantes.

In [33]:
import numpy as np

In [34]:
# Obtengo el maximo valor de datos faltantes y lo guardo en la variable masnan
maxnan_np = np.max(expvida.isnull().sum())
print(maxnan_np)

maxnan_pd = expvida.isnull().sum().max()
print(maxnan_pd)

652
652


In [35]:
datos_no_faltantes = expvida.notnull().values.sum()
datos_faltantes = expvida.isnull().values.sum()
total_datos = datos_no_faltantes + datos_faltantes
print(total_datos)

total = expvida.shape[0] * expvida.shape[1]
print(total)

porcentaje = (datos_faltantes/ total_datos) * 100
print(porcentaje)

64636
64636
3.965282505105514


In [37]:
# Obtengo la cantidad de filas total y divido el maximo valor de datos faltantes por este numero
# y lo multiplico por 100
total_filas = expvida.shape[0]
porc = maxnan_np/total_filas * 100

In [38]:
# Imprimo el maximo porcentaje de datos faltantes. Uso f-strings.
print(f"El porcentaje del máximo valor de datos faltantes es: {porc:.2f}")

El porcentaje del máximo valor de datos faltantes es: 22.19


- str.format()
- Template Class
- f-strings

#### Ahora vamos a realizar algunos calculos y manipulaciones sobre las columnas.

Una función útil para este fin es: `.rename()`. Y podemos utilizarla de la siguiente manera: Si `df` es mi DataFrame y quiero renombrar la columna `columna` como `columna1`, y la columna `columna2` como `columna3`, utilizo la siguiente sintaxis:
  
`df.rename(columns={"columna": "columna1", "columna2": "columna3"}, inplace=True)`

11) Renombra la columna 'Life expectancy ' como 'life_expectancy', ' BMI ' como 'bmi' y ''Measles ' como 'measles'. No olvides de usar `inplace` para modificar el DataFrame original. Luego chequea que el cambio se haya hecho correctamente.

In [39]:
# Renombro columnas del DataFrame
new_columns = {'Life expectancy ' : 'life_expectancy', ' BMI ' : 'bmi', 'Measles ' : 'measles'}
expvida.rename(columns=new_columns, inplace=True)

In [40]:
# Chequeo que se hayan reemplazado correctamente
expvida.columns

Index(['Country', 'Year', 'Status', 'life_expectancy', 'Adult Mortality',
       'infant deaths', 'Alcohol', 'percentage expenditure', 'Hepatitis B',
       'measles', 'bmi', 'under-five deaths ', 'Polio', 'Total expenditure',
       'Diphtheria ', ' HIV/AIDS', 'GDP', 'Population',
       ' thinness  1-19 years', ' thinness 5-9 years',
       'Income composition of resources', 'Schooling'],
      dtype='object')

12) ¿Cual es el promedio de expectativa de vida (Life expectancy) en general?

In [46]:
# Obtengo el promedio de la columna
print(expvida['life_expectancy'].mean())

69.22493169398908


13) ¿Cual es el promedio de expectativa de vida (Life expectancy) por país?

In [47]:
# Obtengo el promedio por pais
#media_pais = expvida.grouphy ()
media_pais = expvida.groupby('Country')[['life_expectancy']].mean()
media_pais

life_expectancy
Country                                            
Afghanistan                                58.19375
Albania                                    75.15625
Algeria                                    73.61875
Angola                                     49.01875
Antigua and Barbuda                        75.05625
...                                             ...
Venezuela (Bolivarian Republic of)         73.38750
Viet Nam                                   74.77500
Yemen                                      63.86250
Zambia                                     53.90625
Zimbabwe                                   50.48750

[193 rows x 1 columns]

14) ¿Cual es la expectativa de vida mas alta y mas baja? Usa la funcion de pandas y la funcion de NumPy y compara el resultado. ¿Que paises tienen esas expectativas de vida?

In [48]:
# Maxima expectativa de vida usando numpy
pais_max_exp = np.max(media_pais)

In [49]:
# Maxima expectativa de vida usando pandas
pais_max_expe = media_pais.max()

In [50]:
# Minima expectativa de vida usando numpy
pais_min_exp = np.min(media_pais)

In [51]:
# Minima expectativa de vida usando pandas
pais_min_expe = media_pais.min()

In [52]:
# Muestro los resultados maximos
print(f'Máxima expectativa de vida usando Numpy: {pais_max_exp}')
print(f'Máxima expectativa de vida usando Pandas: {pais_max_expe}')

Máxima expectativa de vida usando Numpy: 82.5375
Máxima expectativa de vida usando Pandas: life_expectancy    82.5375
dtype: float64


In [53]:
# Muestro los resultados minimos
print(f'Mínima expectativa de vida usando Numpy: {pais_min_exp}')
print(f'Mínima expectativa de vida usando Pandas: {pais_min_expe}')

Mínima expectativa de vida usando Numpy: 46.1125
Mínima expectativa de vida usando Pandas: life_expectancy    46.1125
dtype: float64


In [54]:
# Filtro las columnas para mostrar solo las filas que tengan valor
# igual al maximo
result_max = media_pais[media_pais['life_expectancy'] == pais_max_exp]
result_max

life_expectancy
Country                 
Japan            82.5375

In [55]:
# Filtro las columnas para mostrar solo las filas que tengan valor
# igual al minimo
result_min = media_pais[media_pais['life_expectancy'] == pais_min_exp]
result_min

life_expectancy
Country                      
Sierra Leone          46.1125

15) ¿Cuantas categorias encontramos en la columna "Status"? _**Pista**_: Lee acerca de la función [unique()](http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.unique.html?highlight=unique#pandas.Series.unique)

In [57]:
# Obtengo los valores unicos de la columna status
expvida['Status'].unique()

array(['Developing', 'Developed'], dtype=object)

16) ¿Cuantos paises se ubican en cada una de esas categorias de "Status"? _**Pista**_: Lee acerca de la función [value_counts()](http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.value_counts.html)

In [58]:
# Cuento cuantos valores hay de cada categoria
expvida['Status'].value_counts()

Status
Developing    2426
Developed      512
Name: count, dtype: int64

In [59]:
# ¿Cuántos países se ubican en cada una de las categorías de "Status"?
# Primero agrupo por los valores de la columna "Status"
# Y
# Utilizo la función de Pandas nunique() que cuenta la cantidad de valores únicos en una columna
# En este caso en la columna "Country"

df = expvida.groupby('Status')[['Country']].nunique()
df

Country
Status             
Developed        32
Developing      161

In [60]:
expvida.head()

Country  Year      Status  life_expectancy  Adult Mortality  \
0  Afghanistan  2015  Developing             65.0            263.0   
1  Afghanistan  2014  Developing             59.9            271.0   
2  Afghanistan  2013  Developing             59.9            268.0   
3  Afghanistan  2012  Developing             59.5            272.0   
4  Afghanistan  2011  Developing             59.2            275.0   

   infant deaths  Alcohol  percentage expenditure  Hepatitis B  measles  ...  \
0             62     0.01               71.279624         65.0     1154  ...   
1             64     0.01               73.523582         62.0      492  ...   
2             66     0.01               73.219243         64.0      430  ...   
3             69     0.01               78.184215         67.0     2787  ...   
4             71     0.01                7.097109         68.0     3013  ...   

   Polio  Total expenditure  Diphtheria    HIV/AIDS         GDP  Population  \
0    6.0               8.16         65.0        0.1  584.259210  33736494.0   
1   58.0               8.18         62.0        0.1  612.696514    327582.0   
2   62.0               8.13         64.0        0.1  631.744976  31731688.0   
3   67.0               8.52         67.0        0.1  669.959000   3696958.0   
4   68.0               7.87         68.0        0.1   63.537231   2978599.0   

    thinness  1-19 years   thinness 5-9 years  \
0                   17.2                 17.3   
1                   17.5                 17.5   
2                   17.7                 17.7   
3                   17.9                 18.0   
4                   18.2                 18.2   

   Income composition of resources  Schooling  
0                            0.479       10.1  
1                            0.476       10.0  
2                            0.470        9.9  
3                            0.463        9.8  
4                            0.454        9.5  

[5 rows x 22 columns]

18) ¿Cuantos paises presentaron algun caso de Sarampion (**Measles**) en el 2015? **_Pista_**: Lee acerca de la función [.count()](http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.count.html?highlight=count#pandas.DataFrame.count)

In [61]:
# Cuento cuantas filas tienen valores distintos de cero
data = expvida[expvida['measles'] != 0]
cant_filas = data['measles'].count()
print(f'La cantidad de filas que tienen valores distintos de ceros en la columna "measles" es: {cant_filas}')

La cantidad de filas que tienen valores distintos de ceros en la columna "measles" es: 1955


In [62]:
data_filtered = expvida[(expvida['Year'] == 2015) & (expvida['measles'] > 0)]
data_filtered = data_filtered.drop_duplicates(subset='Country')  #Me aseguro que no haya un país duplicado
data_filtered

Country  Year      Status  life_expectancy  Adult Mortality  \
0     Afghanistan  2015  Developing             65.0            263.0   
32        Algeria  2015  Developing             75.6             19.0   
48         Angola  2015  Developing             52.4            335.0   
96        Armenia  2015  Developing             74.8            118.0   
112     Australia  2015   Developed             82.8             59.0   
...           ...   ...         ...              ...              ...   
2826   Uzbekistan  2015  Developing             69.4            184.0   
2842      Vanuatu  2015  Developing             72.0             13.0   
2874     Viet Nam  2015  Developing             76.0            127.0   
2890        Yemen  2015  Developing             65.7            224.0   
2906       Zambia  2015  Developing             61.8             33.0   

      infant deaths  Alcohol  percentage expenditure  Hepatitis B  measles  \
0                62     0.01               71.279624         65.0     1154   
32               21      NaN                0.000000         95.0       63   
48               66      NaN                0.000000         64.0      118   
96                1      NaN                0.000000         94.0       33   
112               1      NaN                0.000000         93.0       74   
...             ...      ...                     ...          ...      ...   
2826             15      NaN                0.000000         99.0       22   
2842              0      NaN                0.000000         64.0       39   
2874             28      NaN                0.000000         97.0      256   
2890             37      NaN                0.000000         69.0      468   
2906             27      NaN                0.000000          9.0        9   

      ...  Polio  Total expenditure  Diphtheria    HIV/AIDS           GDP  \
0     ...    6.0               8.16         65.0        0.1    584.259210   
32    ...   95.0                NaN         95.0        0.1   4132.762920   
48    ...    7.0                NaN         64.0        1.9   3695.793748   
96    ...   96.0                NaN         94.0        0.1    369.654776   
112   ...   93.0                NaN         93.0        0.1  56554.387600   
...   ...    ...                ...          ...        ...           ...   
2826  ...   99.0                NaN         99.0        0.1   2137.576852   
2842  ...   65.0                NaN         64.0        0.1    285.834140   
2874  ...   97.0                NaN         97.0        0.1           NaN   
2890  ...   63.0                NaN         69.0        0.1           NaN   
2906  ...    9.0                NaN          9.0        4.1   1313.889646   

      Population   thinness  1-19 years   thinness 5-9 years  \
0     33736494.0                   17.2                 17.3   
32    39871528.0                    6.0                  5.8   
48     2785935.0                    8.3                  8.2   
96      291695.0                    2.1                  2.2   
112   23789338.0                    0.6                  0.6   
...          ...                    ...                  ...   
2826    312989.0                    3.0                  3.1   
2842     26463.0                    1.5                  1.4   
2874         NaN                   14.2                 14.5   
2890         NaN                   13.6                 13.4   
2906    161587.0                    6.3                  6.1   

      Income composition of resources  Schooling  
0                               0.479       10.1  
32                              0.743       14.4  
48                              0.531       11.4  
96                              0.741       12.7  
112                             0.937       20.4  
...                               ...        ...  
2826                            0.697       12.1  
2842                            0.598       10.8  
2874                            0.678      

In [63]:
resultado = data_filtered.shape[0]
print(f'La cantidad de países que presentaron algún caso de sarampión (measles) en el 2015 es: {resultado}')

La cantidad de países que presentaron algún caso de sarampión (measles) en el 2015 es: 121
